crates/core/src/db/queries/collections.rs

```diff
@@ -0,0 +1,80 @@
+//! Query utilities for collections.
+
+use anyhow::Context;
+use chrono::{DateTime, Utc};
+use diesel::{
+    pg::Pg,
+    sql_types::{Array, Integer, Nullable, Text, Timestamp},
+    types::ToSql,
+    RunQueryDsl,
+};
+
+use crate::{
+    db::{custom_types::OrderDirection, models::Nft, Connection},
+    error::Result,
+};
+
+/// Query collections ordered by volume
+///
+/// # Errors
+/// returns an error when the underlying queries throw an error
+pub fn by_volume(
+    conn: &Connection,
+    addresses: impl ToSql<Nullable<Array<Text>>, Pg>,
+    order_direction: OrderDirection,
+    start_date: DateTime<Utc>,
+    end_date: DateTime<Utc>,
+    limit: impl ToSql<Integer, Pg>,
+    offset: impl ToSql<Integer, Pg>,
+) -> Result<Vec<Nft>> {
+    diesel::sql_query(make_by_volume_query_string(order_direction))
+        .bind(addresses)
+        .bind::<Timestamp, _>(start_date.naive_utc())
+        .bind::<Timestamp, _>(end_date.naive_utc())
+        .bind(limit)
+        .bind(offset)
+        .load(conn)
+        .context("Failed to load collections by volume")
+}
+
+fn make_by_volume_query_string(order_direction: OrderDirection) -> String {
+    format!(
+        r"
+        WITH volume_table (collection, volume) as (
+            SELECT metadata_collection_keys.collection_address AS collection, SUM(purchases.price) AS volume
+                FROM purchases
+                INNER JOIN metadatas on (purchases.metadata = metadatas.address)
+                INNER JOIN metadata_collection_keys on (metadatas.address = metadata_collection_keys.metadata_address)
+                WHERE
+                    ($1 IS NULL OR metadata_collection_keys.collection_address = ANY($1))
+                    AND purchases.created_at >= $2
+                    AND purchases.created_at <= $3
```

Comment:
Same applies on this query where just want to do volume for SOL based auction houses.

```sql
INNER JOIN auction_houses on (listings.auction_house = auction_houses.address)
  WHERE
    ...
    AND auction_houses.treasury_mint = 'So11111111111111111111111111111111111111112'

Reply:
done!